commerce/domain/promotion/cart/calculator/classes/PromotionCalculatorSample.cls

```diff
@@ -62,39 +67,78 @@ public with sharing class PromotionCalculatorSample extends CartExtension.Promot
         }
     }
 
+    /**
+     * @description Evaluate BOGO promotion (Buy 5 items of qualifying product, get $2 off a unit of target product,
+     * for cart items and apply adjustments.
+     * @param cart Holds details about cart
+     */
+    private static void evaluatePromotionForCartItems(CartExtension.Cart cart) {
+		Integer targetCount = 0;
+        Integer qualifierCount = 0;
+    	Integer adjustmentCount = 0;
+        Iterator<CartExtension.CartItem> ciIter = cart.getCartItems().iterator();
+
+    	if(ciIter.hasNext()) {
+    		qualifierCount = getProductCount(cart, QUALIFIER_PRODUCT_ID);
```

Comment:
We are making time complexity O(n^2) where we don't have to. Why can't we first iterate over cartItems and collect a Map of ProductId to count?

Reply:
`if(cartItemIterator.hasNext())` check on line 84 is an `isEmpty` check. The actual iteration over cartItems happens only once inside `getProductCount()` method. Added comment above to clarify. Thanks!

Reply:
@patricia-bagzai-sf thanks for correcting me. There is no O(n^2) complexity here. I've misread the code.